SLEP006: CalibratedClassifierCV

BenjaminBossan opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
#22893

#### What does this implement/fix? Explain your changes.

This PR adds metadata routing to `CalibratedClassifierCV` (CCV). CCV uses
a subestimator to create (out of sample) probabilities, which are in
turn used to calibrate the probabilities.

The metaestimator uses `sample_weight`. The subestimator may or may not
use `sample_weight` and additional metadata. So far, it was checked if the
subestimator has `sample_weight` in its signature and then they were
routed, otherwise not. This is, however, not always ideal, e.g. when the
subestimator is itself a pipeline (#21134). With routing, this problem disappears.

#### Any other comments?

_The majority of the work here was done pair-programming with @adrinjalali.
Therefore, having a fresh set of eyes to review would be appreciated._

In addition to these changes, the tests in
`test_metaestimator_metadata_routing.py` have been amended to make them
more generic, as right now, they are specific to multioutput.

A current limitation of the generic tests is that `check_recorded_metadata` cannot
be performed for CCV. The reason is that CCV internally creates a slice of the
metadata before passing them to the subestimator. So exact equality fails in this
case. The possibility was discussed to check for exact equality or for the passed
data being a subset; this would work in this case but not in others, e.g. when sample
weights are normalized. Therefore, the solution for now is that in the tests, it can
be declared that this specific metaestimator opts out of `check_recorded_metadata`.

@adrinjalali I still don't use the exact values in `"warns_on"`, please let me know
how to use them exactly. I thought it's easier to discuss this with the code out.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

BenjaminBossan committed:
Make CalibratedClassifierCV work with SLEP006

This PR adds metadata routing to CalibratedClassifierCV (CCV). CCV uses
a subestimator to create (out of sample) probabilities, which are in
turn used to calibrate the probabilities.

The metaestimator uses sample_weight. The subestimator may or may not
use sample_weight and additional metadata. So far, it was checked if the
subestimator has sample_weight in its signature and then they were
routed, otherwise not. This is, however, not always ideal, e.g. when the
subestimator is itself a
pipeline (https://github.com/scikit-learn/scikit-learn/issues/21134).
With routing, this problem disappears.

In addition to these changes, the tests in
test_metaestimator_metadata_routing.py have been amended to make them
more generic, as right now, they are specific to multioutput.

BenjaminBossan committed:
Fix black complaints

BenjaminBossan commented:
_Linting errors seem to be unrelated to this PR_

thomasjpfan commented:
Syncing with `main` should help. The linting error was recently fixed in https://github.com/scikit-learn/scikit-learn/pull/24065

BenjaminBossan commented:
> Syncing with `main` should help. The linting error was recently fixed in #24065

I think since this PR is against `sample-props`, the `sample-props` branch needs to be synced first, right?

adrinjalali commented:
Yeah I'll sync that branch with `main` today.

adrinjalali commented:
`sample-props` branch is now synced with `main`.

BenjaminBossan committed:
Merge branch 'sample-props' into slep006/calibratedclassifiercv

BenjaminBossan commented:
I updated but the linting still fails for unrelated reasons. It also fails on the `sample-props` branch itself because of what looks like changes introduced by the multioutput PR:

https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=45465&view=logs&j=32e2e1bb-a28f-5b18-6cfc-3f01273f5609&t=8a54543f-0728-5134-6642-bedd98e03dd0

I think this PR can thus be reviewed despite the linting errors.

adrinjalali commented on sklearn/tests/test_metaestimators_metadata_routing.py:
```diff
@@ -17,109 +22,279 @@
+
+        if metaestimator.get("preserves_metadata", True):
+            check_recorded_metadata(
+                registry[-1],
```
shouldn't we check all members of registry, to make sure metadata is correctly passed everywhere?

BenjaminBossan replied:
Yes, we can do that, but I'll clear the list between for each method, since there could be cases where the check differs for different methods (at least I think so).

adrinjalali commented on sklearn/tests/test_metaestimators_metadata_routing.py:
```diff
@@ -17,109 +22,279 @@
+            f"{estimator.__class__.__name__}.{method_name}"
         )
+        with pytest.warns(FutureWarning, match=re.escape(warn_msg)):
+            method = getattr(instance, method_name)
+            method(X, y, sample_weight=sample_weight, metadata=metadata)
```
this assumes only `fit` and `partial_fit` route things around, but other methods could do the same, like `transform`, `score` etc.

BenjaminBossan replied:
Do you mean the assumption is implicitly made because of how the method is called? I.e. we need a more generic way to call the method? If so, what needs to be generic: calling with `y`, calling with `sample_weight`?

adrinjalali replied:
the way the method is called is fine, but if the method being called is not `fit` or `partial_fit`, it'll raise an exception that the estimator has not been `fit`.

BenjaminBossan replied:
Hmm, I see. I have no really good solution for this that would work in all cases. Just something OTOH:

```python
# before calling method
if not "fit" in method_name:
    instance.fit(X, y)  # <= would probably still fail on some transformers
...
```

adrinjalali replied:
yeah that would work I think.

BenjaminBossan replied:
Done

adrinjalali commented on sklearn/tests/test_metaestimators_metadata_routing.py:
```diff
@@ -17,109 +22,279 @@
+        instance = cls(**{estimator_name: estimator})
+        # lines below to ensure that there are no warnings
+        with warnings.catch_warnings():
+            warnings.simplefilter("error")
```
we might need to relax this and catch warnings and check that the relevant warnings are not raised. There might be other warnings like convergence warning, etc., which we don't care about in this context.

BenjaminBossan replied:
Okay, I changed the test to record the `FutureWarning`s only, and then check that `"metadata"` is in the message. My reasoning was that we want to avoid raising on other `FutureWarning`s (I guess), so we need to check the message. However, I also didn't want to do an exact check on the whole message. The reason for that is that if someone changes the message but doesn't change this test (which could easily happen), it would pass trivially. I thought checking for `"metadata" is message` would be a good compromise.

adrinjalali commented on sklearn/utils/_metadata_requests.py:
```diff
@@ -1051,6 +1051,10 @@ class RequestMethod:
         function, e.g. ``["sample_weight"]`` if the corresponding method
         accepts it as a metadata.
 
+    validate_keys : bool, default=True
+        Whether to check if the requested parameters fit the actual parameters
+        of the method.
```
note: this was added so that we could simply add an instance of this descriptor to `CheckingClassifier`

adrinjalali reviewed (commented):
Thanks @BenjaminBossan , this looks great!

BenjaminBossan commented on sklearn/tests/test_metaestimators_metadata_routing.py:
```diff
@@ -17,109 +22,279 @@
+    routing_methods = metaestimator["routing_methods"]
+
+    for method_name in routing_methods:
+        if metaestimator.get(method_name, {}).get("warns_on"):
+            # this method is not expected to warn
+            continue
```
So is this the place where the value of `"warns_on"` should be used?

adrinjalali replied:
exactly.

BenjaminBossan replied:
Please take a look at my [latest commit](https://github.com/scikit-learn/scikit-learn/pull/24126/commits/f81389a19b1364746ae9b0f46816ed89f151fbfd) to check if it adequately resolves your concern.

BenjaminBossan committed:
Address reviewer comments by Adrin

- Add __copy__ method to Registry
- Fix parameter docstrings
- Don't repeat metaestimator ids code
- Check_recorded_metadata runs for all registered estimators
- More fine-grained check for warnings, so as _not_ to error on
  unrelated warnings

BenjaminBossan reviewed (commented):
I addressed some of your comments and had some questions on others, please take a look.

adrinjalali commented on sklearn/tests/test_metaestimators_metadata_routing.py:
```diff
@@ -17,109 +22,279 @@
+        "X": X,
+        "y": y_multi,
+        "routing_methods": ["fit", "partial_fit"],
+        "fit": {"warns_on": ["sample_weight"]},
```
it would probably make more sense as a dict as:

```suggestion
        "warns_on": {"fit": ["sample_weight"]},
```

BenjaminBossan replied:
Please take a look at my [latest commit](https://github.com/scikit-learn/scikit-learn/pull/24126/commits/f81389a19b1364746ae9b0f46816ed89f151fbfd) to check if it adequately resolves your concern.

BenjaminBossan committed:
Address reviewer comment: warns_on

Change around structure of the generic test to make more sense. Use the
values to check for specific arguments that should be warned on, instead
of all arguments at once.

adrinjalali commented on sklearn/tests/test_metadata_routing.py:
```diff
@@ -73,14 +73,16 @@ def assert_request_equal(request, dictionary):
     """Utility function to store passed metadata to a method."""
     if not hasattr(obj, "_records"):
-        setattr(obj, "_records", dict())
+        obj._records = {}
```
I don't understand your impulse to change `setattr` :D I also don't use `{}` since to me it's a vague syntax. `{}` is an empty dict, but `{1}` is a `set`.

BenjaminBossan replied:
I think the change I made is more idiomatic than the previous code, which makes it easier to read. `setattr` for me implies that the attribute name is dynamic, which is when you really need it. You wouldn't do `def __init__(self, arg): setattr(self, "arg", arg)`, would you :D 

About the dict: I see your point but I think that `{}` is so common in Python that there is no danger of mistaking it for a set. Furthermore:

```
%timeit _ = {}
21.1 ns ± 0.054 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

%timeit _ = dict()
78.9 ns ± 2.19 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```
;)

All that said, if you want me to revert the change, I'm happy to do that.

adrinjalali commented on sklearn/tests/test_metadata_routing.py:
```diff
@@ -73,14 +73,16 @@ def assert_request_equal(request, dictionary):
     """Check whether the expected metadata is passed to the object's method."""
     records = getattr(obj, "_records", dict()).get(method, dict())
-    assert set(kwargs.keys()) == set(records.keys())
+    # only check keys whose value was explicitly passed
+    expected_keys = {key for key, val in records.items() if val is not None}
+    assert set(kwargs.keys()) == expected_keys
```
Thinking more about this, the change is making the test weaker, and it makes it weaker everywhere.

I think a safer option would be to leave this function as is, and change `record_metadata` to accept a `record_none` arg, which is `True` by default, and in our Consumer estimators in this PR we can set that arg to `False`.

BenjaminBossan replied:
Makes sense, I changed it as you suggested.

adrinjalali commented on sklearn/tests/test_metaestimators_metadata_routing.py:
```diff
@@ -17,109 +22,295 @@
+    warns_on = metaestimator["warns_on"]
+
+    for method_name in routing_methods:
+        if method_name in warns_on:
+            # this method is expected to warn, not raise
+            continue
```
same as the above test, if the method's name is in `warns_on`, it doesn't mean it always warns, it means it warns only for those attributes which are listed there. So we need to test for the other attributes.

BenjaminBossan replied:
You're right, the error test now works analogously to the warning test.

adrinjalali reviewed (commented):
Nice!

BenjaminBossan committed:
Address reviewer comment: recording None

There is now an option for record_metadata to not store None values from
kwargs. This is used in the tests now.

BenjaminBossan committed:
__copy__ has no arguments

There was no error because it's not being used right now.

BenjaminBossan committed:
For expected errors, test each arg seperately

Analogous to the warning test, we want to check each argument in
isolation for the error case.

BenjaminBossan committed:
Fit the metaestimator if not done by method

The could be methods like "score" that require a fitted metaestimator.
Therefore, it is fitted before calling the tested method, except of the
tested method is a fitting method.

Note that right now, this never happens, so in a way that code path is
untested.

BenjaminBossan committed:
Merge branch 'sample-props' into slep006/calibratedclassifiercv

BenjaminBossan commented on sklearn/tests/test_metaestimators_metadata_routing.py:
```diff
@@ -17,109 +23,322 @@
+    """
+
+    def __init__(self, registry=None):
+        self.registry = registry
```
That change would run into trouble when we try to clone the estimator because the parameter was modified during `__init__`.

adrinjalali replied:
I also prefer the way it is. We could add `registry = self.registry or []` at the beginning of all methods instead though. But that's not really shorter.

BenjaminBossan replied:
I left the code as is for now.

thomasjpfan replied:
I see. Okay, let's leave it as is.

adrinjalali commented on sklearn/calibration.py:
```diff
@@ -478,6 +467,33 @@ def predict(self, X):
+            .add(
+                estimator=self._get_estimator(),
+                method_mapping=MethodMapping().add(callee="fit", caller="fit"),
+            )
```
good catch! yes we do. we should add

``` python
.add(splitter=self._get_cv(), method_mapping=MethodMapping().add(callee="split", caller="fit"))
```

and handle splitter args in fit the same way as for `estimator`.

BenjaminBossan replied:
One problem here is that `self._get_cv` would need to get `y` as argument, because `check_cv` takes `y` (and for other stuff). However, that means we cannot call `self._get_cv` in `get_metadata_routing`, since that method does not receive `y`.  

I'm not exactly sure what happens if we just pass `.add(splitter=self.cv, ...)`. It looks like we get a `MetadataRequest(owner=None)` if `self.cv` is, say, an int, is that insufficient?

Tangent: The name "splitter" is arbitrary, right? We could call it anything, as long as we later use the same name when referencing the metadata.

adrinjalali replied:
You're right. I suppose in this case we could actually just pass `self.cv` since the only thing which matters is the metadata requests, and as long as the user doesn't pass a Group*CV there are not requested metadata.

>Tangent: The name "splitter" is arbitrary, right? We could call it anything, as long as we later use the same name when referencing the metadata.

Correct!

BenjaminBossan replied:
Done.

adrinjalali commented on sklearn/tests/test_metadata_routing.py:
```diff
@@ -70,10 +70,18 @@ def assert_request_equal(request, dictionary):
+def record_metadata(obj, method, record_none=True, **kwargs):
+    """Utility function to store passed metadata to a method.
+
+    If record_none is False, kwargs whose values are None are skipped. This is
+    so that checks on keyword arguments whose default was not changed are
+    skipped.
```
We record and check None to make sure if a metadata is not requested, it is not passed, not even as None.

The tests which don't record None, work because the user is not passing any metadata as None, and we need to ignore them cause the default value is `None` and explicitly set in those sub-estimator methods.

Alternatively, we could change the default to `"default"`, and let `record_metadata` know what the _default_ value is, and only ignore that. It might be cleaner.

thomasjpfan replied:
What do you think about always recording `None` and have a flag in `check_recorded_metadata` to switch between the two modes of checking? Concretely:

```python
def check_recorded_metadata(obj, method, strict=True, **kwargs):
    """Check whether the expected metadata is passed to the object's method."""
    records = getattr(obj, "_records", dict()).get(method, dict())

    cmp = operator.eq if strict else operator.le
    assert cmp(set(kwargs.keys()), set(records.keys()))
    for key, value in kwargs.items():
        assert records[key] is value
```

adrinjalali replied:
I kinda prefer the current implementation because it's easier for the test to test exactly what it needs to. What you have here kind of a subset of what the current implementation does, as in, there's no way to test if the router has routed None explicitly or not, and that's something we need to check. I might be missing something here though.

BenjaminBossan committed:
Ignore type checking on CheckingClassifier

Specifically, overriding set_fit_request.

BenjaminBossan committed:
Add routing of groups to splitter

BenjaminBossan committed:
Black formatting

adrinjalali commented on sklearn/calibration.py:
```diff
@@ -496,11 +516,10 @@ def _fit_classifier_calibrator_pair(
     y,
     train,
     test,
-    supports_sw,
     method,
     classes,
     sample_weight=None,
-    **fit_params,
+    fit_params=None,
```
was there a reason for this change?

BenjaminBossan replied:
The call to [`_fit_classifier_calibrator_pair`](https://github.com/BenjaminBossan/scikit-learn/blob/c3a1d67df287569f0335bba9661c03d0fb77fc9c/sklearn/calibration.py#L373-L383) is now analogous to the call directly below to [`cross_val_predict`](https://github.com/BenjaminBossan/scikit-learn/blob/c3a1d67df287569f0335bba9661c03d0fb77fc9c/sklearn/calibration.py#L389-L398). I favored the consistency there but I can also revert if you want.

adrinjalali replied:
That's fair!

adrinjalali reviewed (commented):
Other than this (https://github.com/scikit-learn/scikit-learn/pull/24126#discussion_r945578331), which kinda makes the tests more robust, I'm happy with the PR.

BenjaminBossan committed:
Fix typo

BenjaminBossan committed:
Reviewer request: Change checking of routed data

Give option to not check if the routed metadata are the literal string
"default" (instead of checking for None).

BenjaminBossan committed:
Empty-Commit

thomasjpfan committed:
CI empty-commit

adrinjalali commented:
Why did it not fail when @thomasjpfan pushes? 🤯

BenjaminBossan committed:
Yet another empty commit to try if CI works

BenjaminBossan committed:
New empty commit to try if CI works now

BenjaminBossan committed:
Empty commit after creating CircleCI acct

adrinjalali commented on sklearn/tests/test_calibration.py:
```diff
@@ -1077,3 +1063,21 @@ def test_calibrated_classifier_deprecation_base_estimator(data):
+    X, y = data
+    split_groups = np.array([0, 1] * (len(y) // 2))
+
+    class MyGroupKFold(GroupKFold):
+        """Custom Splitter that checks that the values of groups are correct"""
+
+        def split(self, X, y=None, groups=None):
+            assert (groups == split_groups).all()
+            return super().split(X, y=y, groups=groups)
```
`GroupKFold` raises if `groups` is None anyway, you won't need this custom class to test.

BenjaminBossan replied:
This was more about checking that the correct groups data is being passed. If you think it's not necessary, I'd rather remove it to simplify the test.

adrinjalali replied:
I don't think we have to check for the correctness of the groups. Those mechanisms are tested elsewhere.

adrinjalali reviewed (approved):
I don't mind the custom group splitter either way. LGTM.

thomasjpfan commented on sklearn/tests/test_metadata_routing.py:
```diff
@@ -70,10 +70,22 @@ def assert_request_equal(request, dictionary):
+        kwargs = {
+            key: val
+            for key, val in kwargs.items()
+            if not isinstance(val, str) or (val != "default")
```
@adrinjalali What do you think of https://github.com/scikit-learn/scikit-learn/pull/24126#discussion_r946797598?

I am okay with the current implementation using `"default"` and I do not think this is a blocker.

BenjaminBossan committed:
Don't explicitly check groups values in test

This is already covererd by the general routing tests.

adrinjalali committed:
trigger ci

thomasjpfan reviewed (approved):
LGTM
